nabs/plans.py

```diff
@@ -653,3 +658,131 @@ def daq_a3scan(m1, a1, b1, m2, a2, b2, m3, a3, b3, nsteps):
+                      n_shots, path, snake_like=True):
+    """
+    Scan over two variables in steps simultaneously.
+
+    This is a `nabs` version of ``bluesky``'s built-in
+    `bluesky.plans.list_scan` plan.
+    This scan is designed to be used with 3 motors, the x and y motors that
+    move to a designaget target on a sample wafer, and one scan_motor that can
+    be delay time, laser power or some other motor positions.
+
+    Parameters
+    ----------
+    sample : str
+        The name of the sample we're interested in.
+    detectors : list
+        Objects to read into Python in the scan.
+    x_motor : obj
+        Motor object corresponding to the x axes.
+    y_motor : obj
+        Motor object corresponding to the y axes.
+    scan_motor : obj
+        The motor being scanned. It can be e.g., delay time, laser power, some
+        other motor position, etc.
+    ss : list
+        List of all the points (samples) for the scan_motor to go through.
+    n_shots : int
+        Indicates how many shots should be taken, or how many samples should
+        be scanned on the grid.
+    path : str
+        Path where the sample file is located.
+    snake_like : bool, optional
+        Indicates if the scanning should be following a snake-like pattern.
+        Defaluts to `True`.
+    """
+
+    # TODO: remember what targets have been shot -
+    # display this in a little window - bit mapping?
+    # TODO: if a run gets terminated, remember where to continue
+    # TODO: maybe some validation for the path
+
+    # TODO: take this test _path out and figure out where it goes
+    #  _path = '/home/cristina/workspace/pcdsdevices/samples.yml'
+
+    global _last_index
+    detectors = list(detectors) + [scan_motor]
+
+    rows, columns, last_shot_index, xx, yy = get_sample_info(sample, path)
+    if snake_like:
+        xx = snake_grid_list(np.array(xx).reshape(rows, columns))
+    _last_index = last_shot_index
+    if (last_shot_index + (n_shots * len(ss))) >= len(xx):
+        raise IndexError('The number of n_shots * len(ss): '
+                         f'{last_shot_index + (n_shots * len(ss))} is'
+                         f' bigger than the available samples: {len(xx)}. '
+                         'Please provide a number in range.')
+
+    def inner_scan():
+        global _last_index
+        yield from bps.open_run(md={})
+        for i in range(len(ss)):
+            yield from bps.mv(scan_motor, ss[i])
+            yield from bps.read(scan_motor)
+
+            x_pos = xx[(_last_index + 1):(
+                (_last_index + 1) + n_shots)]
+            y_pos = yy[(_last_index + 1):(
+                (_last_index + 1) + n_shots)]
+            yield from bpp.stub_wrapper(bp.list_scan(detectors, x_motor,
+                                                     x_pos, y_motor, y_pos))
+            _last_index = _last_index + n_shots
+        yield from bps.close_run()
+        update_sample(sample, path, _last_index)
```

Comment:
I am updating the sample file here - assuming that these targets (`x_pos`, `y_pos`) are shot - but I don't know if they have `truly` been shot - is there a better way here to guarantee that the targets have been shot before updating the file to indicate the last target that has been shot?

Reply:
Why don't we know if they have truly been shot? I think if we get here, then at least we've run the sequencer at all of these points, right?

Reply:
Yes correct, I guess we just don't know if the sequencer actually let the pulse go through and hit the target.... but that's already something else - i probably should not worry about that here

Reply:
Oh! I see what you mean. In some cases it might be possible to figure out if the target was hit via live feedback from detector data (it should be a bright hit, not a dark, for example) but maybe this is more complexity than is worthwhile for this.  It's not clear to me exactly what data we'd be having to work with, for example, and it may change from experiment to experiment.

Another thing to think about is interrupted scans; what do we do in that case? Do we mark the first n targets as hit if we got that far?

Reply:
Yeah we want to mark them as shot if they were indeed shot.... somehow
Updating the file a little earlier will probably help with this - and maybe updating after each target is scanned - but this will add some delay that we might not want, and maybe other problems....

I need to do a bit of research about what Bluesky offers in this situation, if we know the last point that has been scanned we can find the index of that point, or we can get the motor position and find the index based on that. Here I would need to know when a scan actually got interrupted.

Reply:
I added a `try/except` block - and I am able to catch when we call `RunEngine.stop()` and `RunEngine.abort()` - the `RunEngine.resume()` should be fine as is - and the `RunEngine.halt()` - I don't know how to deal with this one, since this guy is not doing any cleanup - if we call this we might have to manually update the last_shot_index?